### PR TITLE
`Development`: Add API versioning

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I implemented the changes with a good performance and prevented too many database calls.
- [X] I documented the Java code using JavaDoc style.

### Motivation and Context

For future apps and other external applications, we want to make the API a bit more resilient to breaking changes. Hence, we want to introduce a versioned API.

### Description

I introduced a request mapper that gets injected into the auto configuration. This mapper resolves all endpoints according to their optional versioning annotation. I added extensive JavaDoc and a small documentation. The API prefix `/api` is now part of the request mapper and hence all such annotations were removed from the controllers.
I also added a test that validates the setup of all endpoints in regards to versioning.

Currently, we deactivated the rendering of the API documentation due to performance issues. In the future we want to render the documentation in the CI and publish it similarly to our other documentation: #5932

### Steps for Testing

Changes cover all of the system but not in a functional way. It is expected to be fully backwards-compatible so please try out some stuff you would usually do and check that there are no issues. There are no in-detail tests necessary.

#### Exam Mode Testing

Changes cover all of the system but not in a functional way. It is expected to be fully backwards-compatible so please try out some stuff you would usually do and check that there are no issues. There are no in-detail tests necessary.

#### Code Review
- [x] Review 1
- [x] Review 2
- [x] Review 3
- [x] Review 4
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by ge65cer — COMMENTED** [review]
The core problem seems to be this: https://www.baeldung.com/spring-no-converter-with-preset

This might stem from removing `@RequestMapping("/api")` in all controllers. Maybe that is a lead for debugging?

**Comment by julian-christl** [comment]
> The core problem seems to be this: https://www.baeldung.com/spring-no-converter-with-preset
> 
> This might stem from removing `@RequestMapping("/api")` in all controllers. Maybe that is a lead for debugging?

Yeah, I saw that one too, thanks. Pretty sure it doesn't have anything to do with the annotation but checking this leads me to another finding: It doesn't find any converts at all. Not for `java.time.instant` not for `java.util.ArrayList`. That seems to be the issue.

Let's see... ^^

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java:22]:
```suggestion
       return List.of(range.value());
```
Would this also work?

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:321]:
```suggestion
        return Objects.hash(getRanges(), inRangeCodes);
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Documentation looks good, left some comments about the code

File: src/main/java/de/tum/in/www1/artemis/config/DateTimeFormatConfiguration.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/DateTimeFormatConfiguration.java:3]: It's the one that I have to use, so the warning vanishes. If I use another one it still tells me that the annotations don't match.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:50]: Looks like I resolved a merge conflict wrongly there. Good catch. Would've not affected anything though.

File: src/main/java/de/tum/in/www1/artemis/web/rest/OAuth2JWKSResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/OAuth2JWKSResource.java:25]: Spring sets the content type automatically. This is just a way to override the default behaviour.

This is for example the `/api/exercises/1415/details` call

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java:22]: Sadly no. That's always the thing with int arrays.
But you can mark a stream boxed, then it unpacks the array. Pretty weird.

Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java:22]: What about `Arrays.asList(range.value())` with `java.util.Arrays`? Does that work?

File: src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:133]: In a discussion with @krusche, we mentioned that the thing with constants is that it's really bad for searchability. Hence, I decided to replace them in my PRs, when I'm working on a part containing constants so this will get better over time.

I replaced the other ones.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeRequestCondition.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeRequestCondition.java]: Outdated

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java]: Don't think this works either but I added another solution.

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for addressing our suggestions. Codechanges look good to me

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: I think some examples would be nice to quickly understand what all the version comparison classes output.

**Review by ge65cer — APPROVED** [review]
Tested on TS1 (except exam mode), nice work!
Found an issue when creating exercises, but that seems to be related to Tutorial Groups and not this PR, so I forwarded it to @stefanwaldhauser directly.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: I described each type in `VersionRangeComparionType`. Could you specify what additional documentation you would find helpful and where?

**Review by evgeny1984 — APPROVED** [review]
Tested on TS2, nice work! This is an important feature. Code changes are looking good to me. The documentation provides nice overview and explanation too.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: What I understand so far is we are comparing ranges of version numbers? Is that correct?
Then maybe a few examples would help to understand the code quicker: `Comparing version ranges 1-5  and 2-6 returns xyz`.

File: src/main/java/de/tum/in/www1/artemis/config/DateTimeFormatConfiguration.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/DateTimeFormatConfiguration.java:3]: It was a misconfiguration on my part. I fixed this.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: Do the changes work for you?

Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: Thanks that is great. 👍

**Review by ge65cer — APPROVED** [review]
Approve after feedback was incorporated.

File: src/main/java/de/tum/in/www1/artemis/config/audit/AuditEventConverter.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/audit/AuditEventConverter.java:74]: Spotless says "no"

**Review by Strohgelaender — DISMISSED** [review]
Clarified some questions via Slack, code look good.

**Review by JohannesStoehr — DISMISSED** [review]
Code changes after reviews look good

**Review by evgeny1984 — APPROVED** [review]
Code looks good for me after changes. Reapproved.

**Review by ge65cer — DISMISSED** [review]
Reapprove.

File: docs/dev/api-versioning.rst
Comment (b-fein) [inline docs/dev/api-versioning.rst:13]:
```suggestion
If a range contains two elements, an actual range is specified. ``@VersionRange({1,3})`` specifies that the endpoint accepts versions 1, 2, and 3, assuming that Artemis supports at least until version 3. If a range contains only one element, a lower limit is specified. ``@VersionRange({1})`` specifies that the endpoint accepts all versions starting from 1. ``@VersionRanges`` with no supplied parameters gets interpreted as no annotation. Hence, qualifies for all versions. Other configurations are illegal.
```
Does it actually work with only braces?

**Review by b-fein — APPROVED** [review]
Tested manually:
- SAML2 login works
- GitLab/Jenkins integration works

File: docs/dev/api-versioning.rst
Comment (julian-christl) [reply docs/dev/api-versioning.rst:13]: As a matter of fact, it does not. Good catch.

**Review by b-fein — DISMISSED** [review]
Re-approve. The last change was just a typo in the documentation, so all other reviews are still valid, as well.

**Review by JohannesStoehr — APPROVED** [review]
Documentation change looks good. Good catch Benedict!

**Review by timo07k — DISMISSED** [review]
Tested many different things, there were no issues. Tested on TS1.

**Review by stefanwaldhauser — DISMISSED** [review]
Really cool and nice working feature!

**Review by evgeny1984 — DISMISSED** [review]
Re-approve after fixing the typo in the documentation. The rest functionality is not changed.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:20]: Not sure if I understand your take. The limit (aka lower limit) is from n to infinity.
Plus, that would mean that developers had to specify unintuitive values in the annotations, which is exactly the opposite of what I want.

>Did you consider using the [java.time.temporal.ValueRange](https://docs.oracle.com/javase/8/docs/api/java/time/temporal/ValueRange.html) class?

I did not, but I don't think it's suitable as we don't want to adapt the ranges whenever a new version gets published.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRequestMappingHandlerMapping.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRequestMappingHandlerMapping.java:73]: Not really. It would get unnecessarily complicated. If I wanted to use a ternary operator, which would be necessary here, the parameters would have to be of the same type. But to leave it empty, I would have to use a set; the values of the annotations are given as arrays. That's incompatible.

I could also give the whole condition in a ternary operation, but that would not make it more readable.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java:16]: Not sure if you checked the code out, but VersionRange is an annotation, not a class. I work with lists because arrays are more annoying to me.
I can't implement functionality in an array, as far as I'm aware.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:282]: Changed, let me know what you think of it.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:285]: Saving the compiled pattern into a constant would avoid recompiling multiple times for every request.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:258]:
```suggestion
        if (!path.startsWith("api/") && !path.startsWith("/api/")) {
```
The regex can be avoided easily here.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:285]:
```suggestion
        var pattern = Pattern.compile("^/?api(/v(\\d+))?/.+$");
```
Should `/api/v/…` be valid?

Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:285]: Don't see how this would be valid. The `?` is on `(/v(\\d+))`

Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java]: I think it would be: for `/api/v/…` the `(/v(\\d*))?` the `?`-group part is there since the `/v` part exists, and `\\d*` allows optional digits.

See https://regexr.com/74mr8 first test: it is set to ‘Match none’ but the regex matches.

Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java]: Oh wow. That works differently than I expected. Good catch, then.

Seems like I won't really be able to do that with a regex completely. Probably need to strip it afterwards.

Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java]: Now I get it. Still didn't work as I expected but that doesn't hurt us here.
I completely didn't see your suggested change before.

Yes, that makes sense of course.

The fact that `/api/v/test` is working, is okay. It's just the path `/v/test` without a version.

File: src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:168]: The other regex in `VersionRangesRequestCondition` got changed to `\d+`. Should the patterns in this file also be changed for consistency?
```suggestion
            .antMatchers("/api/{v\\d+}/register").permitAll()
```

**Comment by julian-christl** [comment]
Tests finally run through again. Didn't change anything, no idea what didn't work before but after a clean rebase without merge commits it works like before...

**Review by b-fein — APPROVED** [review]
Code and manual test (local) re-approve.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java
Comment (jpbernius) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java:285]: very good!

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java
Comment (jpbernius) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeService.java:16]: Ok, I was under the wrong impression, that an annotation can offer methods to work with the values it holds. Apparently, that is not the case. So this comment is void.

**Review by jpbernius — APPROVED** [review]
All previous code comments are resolved.

**Review by jpbernius — DISMISSED** [review]

**Review by b-fein — DISMISSED** [review]

**Review by ge65cer — APPROVED** [review]
Re-approve

**Review by JohannesStoehr — APPROVED** [review]
Re-approve

**Comment by krusche** [comment]
I would like to merge this PR only after the PR #5889 was merged. An alternative would be to merge the changes directly into the branch of the PR #5889

**Comment by julian-christl** [comment]
Rebasing on an older base to make it easier to merge into the spring boot 3 branch

**Comment by julian-christl** [comment]
Closed in favour of #6057

**Pull request closed without merging**
